### PR TITLE
CA-234494/CA-234869: introduce default_template field with trigger to update_allowed_operations

### PR DESCRIPTION
**Pull request by mseri**
When attempting the migration of default templates, xapi correctly fails but raises an unclear error.

This fix introduces a preliminary check to stop the cross pool migration with a clearer error when the migration of a default template is detected.

Signed-off-by: Marcello Seri <marcello.seri@citrix.com>

### CODE REVIEW

**Comment by mseri** [comment]
Do not merge. I am still updating the pools to test it.

**Comment by thomassa** [comment]
Do we allow migration of default templates even within a pool? I thought we didn't...

File: ocaml/xapi/xapi_vm_migrate.ml [hunk 1]
```diff
@@ -1045,6 +1045,12 @@ let assert_can_migrate  ~__context ~vm ~dest ~live ~vdi_map ~vif_map ~options =
           "VIF mapping is not allowed for intra-pool migration"]))
 
   | `cross_pool ->
+    (* Prevent Cross-Pool migration of System Templates*)
+    let other_config = Db.VM.get_other_config ~__context ~self:vm in
+    if ((List.mem_assoc Xapi_globs.default_template_key other_config) &&
```
Comment (lindig) [inline ocaml/xapi/xapi_vm_migrate.ml]: You don't need any of the parentheses in the predicate.

```
if List.mem_assoc Xapi_globs.default_template_key other_config &&
  List.assoc Xapi_globs.default_template_key other_config="true" then 
    ..  ;
```

**Comment by mseri** [comment]
@thomassa do you know where I can find a design/specification for that? I could move the check to the beginning of the function in that case

**Comment by thomassa** [comment]
No, but migrating a template within a pool is essentially meaningless.
It is only a database entry, and can never be a running VM, so within a pool it is not on any particular host.
Copying (rather than migrating) a template from one pool to another could make sense, but probably not for system templates.

**Comment by mseri** [comment]
I understand, I think the main issue is that XenCenter was trying to migrate the template (the copy works without problems) and presenting the wrong wizard. But even in that case it made sense for us to provide an understandable message. Unfortunately I could reproduce the issue only once so for the moment I cannot see what happens if I try a migration within the pool.

**Comment by mseri** [comment]
As an aside, the CLI does already the right thing, reporting e.g.
```
The operation attempted is not valid for a template VM
vm: f873abe0-b138-4995-8f6f-498b423d234d (Oracle Linux 7)
```

I should probably use the same error `vm_is_template` instead, thoughts?

**Comment by thomassa** [comment]
I think that would make sense.

**Comment by mseri** [comment]
I modified it to prevent any template migration. I will test it on Monday morning to make sure that cli and api are doing the correct thing.





On Fri, Dec 2, 2016 at 5:52 PM +0000, "thomassa" <notifications@github.com<mailto:notifications@github.com>> wrote:

I think that would make sense.

-

**Comment by mseri** [comment]
I duscussed the issue with @jonludlam and I am going to revert to the previous change. The migration is used also for storage migration and not just memory, so we should be able to migrate templates but not default templates.

While testing the fix with the xmlrpc api, I have notice a discrepancy in the checks of `pool_migrate` and `assert_vm_can_migrate` (used by `migrate_send`), I am also going to check why this happens and make it more uniform.

**Comment by mseri** [comment]
There have been more discussions related to this, minor, issue. The issue is rare and appears when the allowed operations for default templates are not updated, this happens only the first time the template is used and is due to the way these templates are generated at the moment. The fix that seems more reasonable is to make a new `is_default_template` field that calls `update_allowed_operation` when set.

**Comment by mseri** [comment]
In addition to unit tests and quicktest, I tested it by setting all the templates to default templates on two new pools, attempting copy and migrations inter and intra pool from API and XenCenter, creating new VMs from the default templates, doing some copy/migration/snapshot operations.

File: ocaml/idl/datamodel.ml [hunk 2]
```diff
@@ -7611,6 +7624,7 @@ let vm =
 
          field ~ty:Int "user_version" "Creators of VMs and templates may store version information here.";
          field ~effect:true ~ty:Bool "is_a_template" "true if this is a template. Template VMs can never be started, they are used only for cloning other VMs";
+         field ~ty:Bool ~default_value:(Some (VBool false)) ~qualifier:DynamicRO ~writer_roles:_R_POOL_ADMIN ~lifecycle:[Published, rel_ely, "Identifies XenServer default templates"] "is_default_template" "true if this is a default template. Deafult template VMs can never be started or migrated, they are used only for cloning other VMs";
```
Comment (lindig) [inline ocaml/idl/datamodel.ml]: Typo: `Deafult template VMs can never`

File: ocaml/xapi/import.ml [hunk 3]
```diff
@@ -293,9 +293,12 @@ module VM : HandlerTools = struct
 
   let precheck __context config rpc session_id state x =
     let vm_record = API.Legacy.From.vM_t "" x.snapshot in
-    if vm_record.API.vM_is_a_template
-    && (List.mem_assoc Xapi_globs.default_template_key vm_record.API.vM_other_config)
-    && ((List.assoc Xapi_globs.default_template_key vm_record.API.vM_other_config) = "true")
+    let is_default_template = vm_record.API.vM_is_default_template || (
```
Comment (lindig) [inline ocaml/xapi/import.ml]: This `is_default_template` is almost identical to the code in `export.ml`. How about defining the function in one of the helper modules?

File: ocaml/idl/datamodel.ml (hunk 2 above)
Comment (mseri) [inline ocaml/idl/datamodel.ml]: Well spotted!

File: ocaml/xapi/import.ml (hunk 3 above)
Comment (mseri) [inline ocaml/xapi/import.ml]: I kept them separate hoping that we decide to deprecate the use of `other_config` for the default templates and we can get rid of the second part of the check.

**Comment by lindig** [comment]
I think this is now fine to go.

**Comment by mseri** [comment]
I squash the commits

**Pull request merged**